source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+    , global_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
+  {
+    hdf5_reference = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Dclose(*pointer);
+      delete pointer;
+    });
+    dataspace      = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Sclose(*pointer);
+      delete pointer;
+    });
+
+    // rank_ret can take a negative value if the functions
+    // H5Sget_simple_extent_ndims and H5Sget_simple_extent_dims fail. rank is
+    // unsigned int therefore it can not be used to store the return value of
+    // H5Sget_simple_extent_ndims and H5Sget_simple_extent_dims
+    int rank_ret;
+
+    *hdf5_reference = H5Dopen2(parent_group_id, name.data(), H5P_DEFAULT);
+    Assert(*hdf5_reference >= 0, ExcMessage("Error at H5Dopen2"));
+    *dataspace = H5Dget_space(*hdf5_reference);
+    Assert(*dataspace >= 0, ExcMessage("Error at H5Dget_space"));
+    rank_ret = H5Sget_simple_extent_ndims(*dataspace);
+    Assert(rank_ret >= 0, ExcInternalError());
+    rank          = rank_ret;
+    hsize_t *dims = (hsize_t *)malloc(rank * sizeof(hsize_t));
+    rank_ret      = H5Sget_simple_extent_dims(*dataspace, dims, NULL);
+    AssertDimension(rank_ret, static_cast<int>(rank));
+    dimensions.assign(dims, dims + rank);
+    free(dims);
+
+    size = 1;
+    for (auto &&dimension : dimensions)
+      {
+        size *= dimension;
+      }
+  }
+
+
+
+  DataSet::DataSet(const std::string &         name,
+                   const hid_t &               parent_group_id,
+                   const std::vector<hsize_t> &dimensions,
+                   std::shared_ptr<hid_t>      t_type,
+                   const bool                  mpi)
+    : HDF5Object(name, mpi)
+    , rank(dimensions.size())
+    , dimensions(dimensions)
+    , query_io_mode(false)
+    , io_mode(H5D_MPIO_NO_COLLECTIVE)
+    , local_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
+    , global_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
+  {
+    hdf5_reference = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Dclose(*pointer);
+      delete pointer;
+    });
+    dataspace      = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Sclose(*pointer);
+      delete pointer;
+    });
+
+    *dataspace = H5Screate_simple(rank, dimensions.data(), NULL);
+    Assert(*dataspace >= 0, ExcMessage("Error at H5Screate_simple"));
+
+    *hdf5_reference = H5Dcreate2(parent_group_id,
+                                 name.data(),
+                                 *t_type,
+                                 *dataspace,
+                                 H5P_DEFAULT,
+                                 H5P_DEFAULT,
+                                 H5P_DEFAULT);
+    Assert(*hdf5_reference >= 0, ExcMessage("Error at H5Dcreate2"));
+
+    size = 1;
+    for (auto &&dimension : dimensions)
```

Comment:
Here and maybe elsewhere: it's not wrong to use rvalue references as long as you have just scalars in your arrays. But it raises questions about whether any data is moved. I would just use regular references in `for` loops.